### PR TITLE
gps2tsdb Testing

### PR DESCRIPTION
**Pull request by aaron97neu**
Verification of gps2tsdb container. Contains testing container that provides it's own dbus. Reads file formatted in output from `gpspipe -w` and then checks that each point is properly recorded into the database

TODOs:
- [x] How to shutdown containers when testing is finished and communicate result to another script
- [x] Clear database before starting
- [x] Move github actions out of `build-and-push-services` and into it's own test workflow
- [x] General code cleanup
- [x] discuss where to put testing containers
- [x] add new files to codeowners

### CODE REVIEW

**Comment by aaron97neu** [comment]
Added zmq nng support. @abalmos and @facastiblancor if you guys want to see zeromq and nng side by side it should be fairly usable

Not perfect yet, will improve readability once we choose a messaging lib

**Comment by aaron97neu** [comment]
Finished all objectives here. This should be ready for review

**Review by facastiblancor — APPROVED** [review]
Looks good to me.

File: services/gps_verify/docker-compose-example.yml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+    build:
+      context: ../gps_verify
+    environment:
+      db_user: "avena"
+      db_host: "postgres"
+      db_database: "avena"
+      db_password: "password"
+      db_port: "5432"
+    volumes:
+      - ../gps_verify/dbus:/var/run/dbus:rw
```
Comment (abalmos) [inline services/gps_verify/docker-compose-example.yml:13]: Can we move this file to a Docker volume, rather then exposing it to the host?

**Review by abalmos — APPROVED** [review]
As it, I am okay with it, but I do have questions (and a few comments left behind):

Is there a reason we are exposing `gps_replay` and `gps_verify` as separate services, rather then just as the tests for `gps2tsdb`?

If it is for reuse purposes, I think that might be okay, but I suggest we move these test services into another location. I feel like `services` should be for deployable things.

File: services/gps2tsdb/gps.py [hunk 2]
```diff
@@ -28,7 +33,28 @@ def fix(*args):
...
+    if( tries < maxtries):
+      print("Database connection attempt", tries, "failed. Database may still be starting. Sleeping", sleeptime, "s and trying again")
+      print(e)
+      tries = tries + 1
+      sleep(sleeptime)
+    else:
+      print("FATAL: Could not connect to db after", tries, "tries. Exiting")
+      sys.exit(-1)
+  else:
+    db_connected = True
```
[the remark this answers is not in the document]
Comment (aaron97neu) [reply services/gps2tsdb/gps.py:57]: Not that I could tell. The `wait-for-postgres.sh` example linked below seems to be how it is typically done but we do not have psql in the container atm and it has a surprising amount of dependencies. Decided this was the lightest way to do it. Not a huge fan of it

https://docs.docker.com/compose/startup-order/

File: services/gps_verify/docker-compose-example.yml (hunk 1 above)
Comment (aaron97neu) [reply services/gps_verify/docker-compose-example.yml:13]: Good idea, will give it a shot

File: services/gps_replay/gps_replay.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+# Get synchronization from subscribers
+print('awaiting', subs_expected, 'subs')
+subscribers = 0
+while subscribers < subs_expected:
+    # wait for synchronization request
+    msg = zmq_sync.recv()
+    # send synchronization reply
+    zmq_sync.send(b'')
+    subscribers += 1
+    print('+1 subscriber (%i/%i)' % (subscribers, subs_expected))
```
[the remark this answers is not in the document]
Comment (aaron97neu) [reply services/gps_replay/gps_replay.py:90]: Agreed. We can talk if we want to migrate it to NATS before merging tomorrow. This PR was opened before we investigated NATS at all and I had not thought to update it

**Comment by aaron97neu** [comment]
>Is there a reason we are exposing gps_replay and gps_verify as separate services, rather then just as the tests for gps2tsdb?
>
>If it is for reuse purposes, I think that might be okay, but I suggest we move these test services into another location. I feel like services should be for deployable things.

@abalmos agreed. I meant to ask about this in a meeting but I totally forgot. I originally put in it `services` because our build environment builds each folder here as a container. I was thinking we could move it to a `test` folder alongside services, or within the `gps2tsdb` folder. 

I'm happy to put them anywhere but we can discuss it tomorrow 

As for reuse I would imagine any other container that consumes gps points from gpsd would reuse the gps point generation container for it's own integration testing. I was modeling this system off of this diagram we drew in our Nov 2nd meeting:
image

**Comment by aaron97neu** [comment]
@abalmos Updated with the NATS messaging system. Take a look when you get a second

File: services/gps2tsdb/test/gps_verify/gps_verify.py [hunk 4]
```diff
@@ -1,75 +1,104 @@
...
-zmq_sync = zmq_ctx.socket(zmq.REQ)
-zmq_sync.connect('tcp://gps_replay:5562')
+    # Debugging
+    #print("NATS type:", type(nats_gps_point), "\nmessage:", nats_gps_point)
+    
+    # Check if this is a control message signifying the end of the transmission
+    if nats_gps_point == "END": 
+      print('End control message received. No more gps points to check')
+      nonlocal fin
+      fin.set_result("END")
```
Comment (abalmos) [inline services/gps2tsdb/test/gps_verify/gps_verify.py:67]: Maybe this is the way for python? I'm more used to message passing based flows, but that may not be "pythonic"?

**Review by abalmos — APPROVED** [review]
Seems like asyncio noise is mostly around just doing it asyncio's way. Likely will need that type of structure for most things given the event driven-ness of our services.

Might check out the gist I shared. I'm not sure if this PRs method of using a Future to sync to the "logical threads" or using messing passing to put the logic into one work function is more "pythonic". Do you know?

File: services/gps2tsdb/test/gps_replay/gps_replay.py [hunk 5]
```diff
@@ -6,203 +6,205 @@
...
-# Get synchronization from subscribers
-print('awaiting', subs_expected, 'subs')
-subscribers = 0
-while subscribers < subs_expected:
-    # wait for synchronization request
-    msg = zmq_sync.recv()
-    # send synchronization reply
-    zmq_sync.send(b'')
-    subscribers += 1
-    print('+1 subscriber (%i/%i)' % (subscribers, subs_expected))
```
[the remark this answers is not in the document]
Comment (aaron97neu) [reply services/gps2tsdb/test/gps_replay/gps_replay.py]: Yes and no. The subscriber still has to arrive first and I started by trying to emulate that in NATS but gave up and just put a sleep in the publisher to ensure the subscriber arrives first

File: services/gps2tsdb/test/gps_verify/gps_verify.py [hunk 6]
```diff
@@ -1,75 +1,104 @@
...
-db.run('TRUNCATE gps;')
+  # Try to connect to the DB. It may be starting up so we should try a few timees
+  # before failing. Currently trying every second 60 times
+  tries = 0
+  maxtries = 60
+  sleeptime = 1
+  db_connected = False
+  while not db_connected:
+    try:
+      db = postgres.Postgres(url=connectionurl)
```
[the remark this answers is not in the document]
Comment (aaron97neu) [reply services/gps2tsdb/test/gps_verify/gps_verify.py:27]: oooh I did not know about that lib. Very fancy. Def for use in future, esp for production containers

File: services/gps2tsdb/test/gps_verify/gps_verify.py [hunk 7]
```diff
@@ -1,75 +1,104 @@
...
+  print('Truncating GPS database')
+  db.run('TRUNCATE gps;')
+    
+  print('Setting up NATS')
+  # Connect to NATS server 
+  nc = NATS()
+  await nc.connect("nats://nats:4222")
+  
+  # Notifies the non-callback function that we are finished
+  fin = asyncio.Future()
```
[the remark this answers is not in the document]
Comment (aaron97neu) [reply services/gps2tsdb/test/gps_verify/gps_verify.py:55]: Prob not pythonic. Pulled from one of NATS's examples. Real message passing is prob the better way
https://docs.nats.io/developing-with-nats/receiving/async

File: services/gps2tsdb/test/gps_replay/gps_replay.py [hunk 8]
```diff
@@ -6,203 +6,205 @@
...
+            
+                # Don't overpower gps2tsdb
+                # Basically if this is not here, gps_verify will check for the point before
+                # gps2tsdb has the chance to enter it into the db
+                sleep(.1)
+    
+                # Publish GPS point to zeromq
+                print('Publishing gps data')
+                await nc.publish("gps", json.dumps(point).encode());
+                await nc.flush(1)
```
[the remark this answers is not in the document]
Comment (aaron97neu) [reply services/gps2tsdb/test/gps_replay/gps_replay.py:194]: Was needed at first for debugging but isn't as necessary, but makes the logs much prettier. The different between this
```
gps2tsdb_1    | Time:  1604351107.0 	Lat:  40.4327735 	Lng:  -86.914165333
gps2tsdb_1    | Inserting lat and lng for timestamp  1604351107.0
gps2tsdb_1    | Finished inserting lat and lng for timestamp  1604351107.0
gps_replay_1  | Tuple: (1604351107.0, 3, 0.005, 40.4327735, -86.914165333, 24.89, 194.6, 50.14, nan, nan, 0.527, 28.72, 0.0, 100.28, 'fake-data-verification')
gps_replay_1  | Publishing gps data
gps2tsdb_1    | Time:  1604351108.0 	Lat:  40.4327705 	Lng:  -86.914167
gps2tsdb_1    | Inserting lat and lng for timestamp  1604351108.0
gps2tsdb_1    | Finished inserting lat and lng for timestamp  1604351108.0
gps_replay_1  | Tuple: (1604351108.0, 3, 0.005, 40.4327705, -86.914167, 24.89, 195.5, 50.14, nan, nan, 0.309, 28.72, 0.9, 100.28, 'fake-data-verification')
gps_replay_1  | Publishing gps data
gps_verify_1  | Verifying that nats 2020-10-26T19:47:55.000Z  is in the database
gps_verify_1  | nats timestamp 2020-10-26T19:47:55.000Z successfully entered into the database
gps_verify_1  | Verifying that nats 2020-10-26T19:47:56.000Z  is in the database
gps_verify_1  | nats timestamp 2020-10-26T19:47:56.000Z successfully entered into the database
gps_verify_1  | Verifying that nats 2020-10-26T19:47:57.000Z  is in the database
gps_verify_1  | nats timestamp 2020-10-26T19:47:57.000Z successfully entered into the database
gps_verify_1  | Verifying that nats 2020-10-26T19:47:58.000Z  is in the database
gps_verify_1  | nats timestamp 2020-10-26T19:47:58.000Z successfully entered into the database
gps_verify_1  | Verifying that nats 2020-10-26T19:47:59.000Z  is in the database
gps_verify_1  | nats timestamp 2020-10-26T19:47:59.000Z successfully entered into the database
gps_verify_1  | Verifying that nats 2020-10-26T19:48:00.000Z  is in the database
gps_verify_1  | nats timestamp 2020-10-26T19:48:00.000Z successfully entered into the database
```
and this: (Not that the verify messages are interlaced within the gps point that was sent)
```
gps_verify_1  | Verifying that nats 2020-11-02T21:05:04.000Z  is in the database
gps_verify_1  | nats timestamp 2020-11-02T21:05:04.000Z successfully entered into the database
gps2tsdb_1    | Time:  1604351105.0 	Lat:  40.432775833 	Lng:  -86.914151667
gps2tsdb_1    | Inserting lat and lng for timestamp  1604351105.0
gps2tsdb_1    | Finished inserting lat and lng for timestamp  1604351105.0
gps_replay_1  | Tuple: (1604351105.0, 3, 0.005, 40.432775833, -86.914151667, 24.89, 194.6, 50.14, nan, nan, 0.34, 28.72, -0.2, 100.28, 'fake-data-verification')
gps_replay_1  | Publishing gps data
gps_verify_1  | Verifying that nats 2020-11-02T21:05:05.000Z  is in the database
gps_verify_1  | nats timestamp 2020-11-02T21:05:05.000Z successfully entered into the database
gps2tsdb_1    | Time:  1604351106.0 	Lat:  40.432774167 	Lng:  -86.914155333
gps2tsdb_1    | Inserting lat and lng for timestamp  1604351106.0
gps2tsdb_1    | Finished inserting lat and lng for timestamp  1604351106.0
gps_replay_1  | Tuple: (1604351106.0, 3, 0.005, 40.432774167, -86.914155333, 24.89, 194.6, 50.14, nan, nan, 0.211, 28.72, 0.0, 100.28, 'fake-data-verification')
gps_replay_1  | Publishing gps data
gps_verify_1  | Verifying that nats 2020-11-02T21:05:06.000Z  is in the database
gps_verify_1  | nats timestamp 2020-11-02T21:05:06.000Z successfully entered into the database
gps2tsdb_1    | Time:  1604351107.0 	Lat:  40.4327735 	Lng:  -86.914165333
gps2tsdb_1    | Inserting lat and lng for timestamp  1604351107.0
gps2tsdb_1    | Finished inserting lat and lng for timestamp  1604351107.0
gps_replay_1  | Tuple: (1604351107.0, 3, 0.005, 40.4327735, -86.914165333, 24.89, 194.6, 50.14, nan, nan, 0.527, 28.72, 0.0, 100.28, 'fake-data-verification')
gps_replay_1  | Publishing gps data
gps_verify_1  | Verifying that nats 2020-11-02T21:05:07.000Z  is in the database
gps_verify_1  | nats timestamp 2020-11-02T21:05:07.000Z successfully entered into the database
gps2tsdb_1    | Time:  1604351108.0 	Lat:  40.4327705 	Lng:  -86.914167
gps2tsdb_1    | Inserting lat and lng for timestamp  1604351108.0
gps2tsdb_1    | Finished inserting lat and lng for timestamp  1604351108.0
gps_replay_1  | Tuple: (1604351108.0, 3, 0.005, 40.4327705, -86.914167, 24.89, 195.5, 50.14, nan, nan, 0.309, 28.72, 0.9, 100.28, 'fake-data-verification')
gps_replay_1  | Publishing gps data
gps_verify_1  | Verifying that nats 2020-11-02T21:05:08.000Z  is in the database
gps_verify_1  | nats timestamp 2020-11-02T21:05:08.000Z successfully entered into the database
```

File: services/gps2tsdb/test/gps_replay/gps_replay.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (aaron97neu) [reply services/gps2tsdb/test/gps_replay/gps_replay.py]: The reason I left out this functionality is that in NATS (and zmq too really), if I were to send a reqest-response, the other container would need to already be subscribed to get the request so it really doesn't solve the synchronization problem as we still need to guarantee that one container is first. There may be a solution but I'd also argue that tests should err on the side of simplicity even if it means a few extra seconds of run time

[the remark this answers is not in the document]
Comment (abalmos) [reply services/gps2tsdb/test/gps_replay/gps_replay.py]: Interesting, we will run into this problem in practice too.

I wonder if our services should all have a "ping" type req/rep setup. So a service can loop on trying to send a ping to a downstream service to check it's status. Could be used as part of the health checks too?

**Pull request merged**